File: e2e_walkthrough.js
const { chromium } = require('/opt/homebrew/lib/node_modules/playwright');

const BASE = 'http://localhost:3010';
const results = [];

function log(cat, name, passed, detail) {
  results.push({cat, name, passed, detail});
  const icon = passed ? '✅' : '❌';
  let msg = `  ${icon} ${cat}: ${name}`;
  if (detail) msg += ` (${detail})`;
  console.log(msg);
}

(async () => {
  const browser = await chromium.launch({headless: true});
  const ctx = await browser.newContext({viewport: {width: 1280, height: 800}});
  const page = await ctx.newPage();

  const consoleErrors = [];
  const networkErrors = [];
  page.on('console', msg => { if (msg.type() === 'error') consoleErrors.push(msg.text()); });
  page.on('requestfailed', req => networkErrors.push(`${req.method()} ${req.url()}`));

  console.log('='.repeat(60));
  console.log('Knowledge OS E2E Walkthrough');
  console.log('='.repeat(60));

  // ==========================================
  // 1. LOGIN PAGE
  // ==========================================
  console.log('\n[1] Login Page');
  const resp = await page.goto(BASE, {waitUntil: 'networkidle', timeout: 15000});
  log('Landing', 'Page loads', resp.status() === 200);
  log('Landing', 'Title', (await page.title()) === 'Knowledge OS');
  log('Login', 'Email input (#login-email)', await page.locator('#login-email').count() > 0);
  log('Login', 'Password input (#login-password)', await page.locator('#login-password').count() > 0);
  log('Login', 'Sign in button', await page.locator('button:has-text("Sign in")').count() > 0);
  log('Login', 'Sign up button', await page.locator('button:has-text("Sign up")').count() > 0);
  log('Login', 'Forgot password', await page.locator('a:has-text("Forgot password")').count() > 0);

  // ==========================================
  // 2. REGISTER
  // ==========================================
  console.log('\n[2] Registration');
  await page.locator('button:has-text("Sign up")').click();
  await page.waitForTimeout(1000);
  log('Register', 'Register form visible', await page.locator('#register-email').count() > 0);

  await page.locator('#register-email').fill('e2ewalk@test.com');
  await page.locator('#register-username').fill('e2ewalk');
  await page.locator('#register-password').fill('WalkPass123!');
  await page.locator('#register-confirm-password').fill('WalkPass123!');

  await page.locator('button[type="submit"]').click();
  await page.waitForTimeout(3000);

  const registered = page.url() === BASE || page.url() === BASE + '/';
  log('Register', 'Success', registered, `url=${page.url()}`);

  if (!registered) {
    const errs = await page.evaluate(() => {
      return [...document.querySelectorAll('[class*="destructive"], [role="alert"]')].map(e => e.textContent?.trim()).filter(Boolean);
    });
    if (errs.length) log('Register', 'Error', false, errs[0]?.slice(0,100));
  }

  // ==========================================
  // 3. LOGGED-IN PAGES
  // ==========================================
  const isApp = registered || !page.url().includes('/login');

  if (!isApp) {
    // Try login
    console.log('\n[3] Login (fallback)');
    if (await page.locator('button:has-text("Sign in")').count() > 0) {
      await page.locator('button:has-text("Sign in")').click();
      await page.waitForTimeout(500);
    }
    await page.locator('#login-email').fill('e2ewalk@test.com');
    await page.locator('#login-password').fill('WalkPass123!');
    await page.locator('button[type="submit"]').click();
    await page.waitForTimeout(3000);
  }

  const loggedIn = !page.url().includes('/login') && !page.url().includes('/register');
  log('App', 'Logged in', loggedIn, `url=${page.url()}`);

  if (loggedIn) {
    // ==========================================
    // 4. OUTLINER
    // ==========================================
    console.log('\n[4] Outliner');
    const outlinerText = await page.evaluate(() => document.body.innerText);
    log('Outliner', 'Page renders', outlinerText.length > 100, `${outlinerText.length} chars`);

    const hasSidebar = await page.locator('nav, aside, [class*="sidebar"]').count() > 0;
    log('Outliner', 'Sidebar present', hasSidebar);

    const hasNewBtn = await page.locator('button:has-text("New"), button:has-text("Create"), button:has-text("+")').count() > 0;
    log('Outliner', 'New/Create button', hasNewBtn);

    if (hasNewBtn) {
      await page.locator('button:has-text("New"), button:has-text("Create"), button:has-text("+")').nth(0).click();
      await page.waitForTimeout(1000);

      const typeOpt = page.locator('text=Note, text=Page, text=Task');
      if (await typeOpt.count() > 0) {
        await typeOpt.nth(0).click();
        await page.waitForTimeout(500);

        const titleInput = page.locator('input[placeholder*="title"], input[name="title"]');
        if (await titleInput.count() > 0) {
          await titleInput.first.fill('E2E Walkthrough Note');

          const submitBtn = page.locator('button:has-text("Create"), button:has-text("Save"), button[type="submit"]');
          if (await submitBtn.count() > 0) {
            await submitBtn.last.click();
            await page.waitForTimeout(2000);
            log('Outliner', 'Created object', true);
          }
        }
      }
    }

    // ==========================================
    // 5. SEARCH
    // ==========================================
    console.log('\n[5] Search');
    await page.goto(`${BASE}/search`, {waitUntil: 'networkidle', timeout: 10000});
    await page.waitForTimeout(1000);
    log('Search', 'Page loads', !page.url().includes('login'));

    const searchInput = page.locator('input[type="search"], input[placeholder*="search"], input[name="q"]');
    if (await searchInput.count() > 0) {
      await searchInput.first.fill('walkthrough');
      await page.waitForTimeout(2000);
      log('Search', 'Search executed', true);
    } else {
      log('Search', 'Search input', false, 'not found');
    }

    // ==========================================
    // 6. TASKS
    // ==========================================
    console.log('\n[6] Tasks');
    await page.goto(`${BASE}/tasks`, {waitUntil: 'networkidle', timeout: 10000});
    await page.waitForTimeout(1000);
    log('Tasks', 'Page loads', !page.url().includes('login'));
    if (!page.url().includes('login')) {
      const text = await page.evaluate(() => document.body.innerText);
      log('Tasks', 'Has task UI', ['task', 'priority', 'status'].some(kw => text.toLowerCase().includes(kw)));
    }

    // ==========================================
    // 7. AGENTS
    // ==========================================
    console.log('\n[7] Agents');
    await page.goto(`${BASE}/agents`, {waitUntil: 'networkidle', timeout: 10000});
    await page.waitForTimeout(1000);
    log('Agents', 'Page loads', !page.url().includes('login'));
    if (!page.url().includes('login')) {
      const text = await page.evaluate(() => document.body.innerText);
      log('Agents', 'Has agent UI', ['agent', 'chat', 'status'].some(kw => text.toLowerCase().includes(kw)));
    }

    // ==========================================
    // 8. FILES
    // ==========================================
    console.log('\n[8] Files');
    await page.goto(`${BASE}/files`, {waitUntil: 'networkidle', timeout: 10000});
    await page.waitForTimeout(1000);
    log('Files', 'Page loads', !page.url().includes('login'));
    if (!page.url().includes('login')) {
      const text = await page.evaluate(() => document.body.innerText);
      log('Files', 'Has file UI', ['file', 'folder', 'watch'].some(kw => text.toLowerCase().includes(kw)));
    }

    // ==========================================
    // 9. SETTINGS
    // ==========================================
    console.log('\n[9] Settings');
    await page.goto(`${BASE}/settings`, {waitUntil: 'networkidle', timeout: 10000});
    await page.waitForTimeout(1000);
    log('Settings', 'Page loads', !page.url().includes('login'));
    if (!page.url().includes('login')) {
      const text = await page.evaluate(() => document.body.innerText);
      log('Settings', 'Has settings UI', ['setting', 'backup', 'openclaw'].some(kw => text.toLowerCase().includes(kw)));
    }

    // ==========================================
    // 10. NAVIGATION
    // ==========================================
    console.log('\n[10] Navigation');
    await page.goto(BASE, {waitUntil: 'networkidle', timeout: 10000});
    const navLinks = page.locator('nav a, aside a, [class*="sidebar"] a');
    log('Navigation', `Nav links (${await navLinks.count()})`, await navLinks.count() >= 3);
    await page.screenshot({path: 'e2e-screenshot-final.png', fullPage: true});
    log('Navigation', 'Screenshot', true, 'e2e-screenshot-final.png');
  } else {
    console.log('\n[4-10] Skipped — could not log in');
    log('App', 'All pages skipped', false, 'login failed');
  }

  // ==========================================
  // SUMMARY
  // ==========================================
  console.log('\n' + '='.repeat(60));
  const passed = results.filter(r => r.passed).length;
  const failed = results.filter(r => !r.passed).length;
  console.log(`Results: ${passed}/${results.length} passed, ${failed} failed`);
  console.log('='.repeat(60));

  if (failed > 0) {
    console.log('\nFailures:');
    results.filter(r => !r.passed).forEach(r => console.log(`  ❌ ${r.cat}: ${r.name} — ${r.detail || ''}`));
  }

  if (consoleErrors.length > 0) {
    console.log(`\n⚠️  Console errors (${consoleErrors.length}):`);
    consoleErrors.slice(0, 5).forEach(e => console.log(`  ${e.slice(0, 150)}`));
  }

  if (networkErrors.length > 0) {
    console.log(`\n⚠️  Network errors (${networkErrors.length}):`);
    networkErrors.slice(0, 5).forEach(e => console.log(`  ${e.slice(0, 150)}`));
  }

  await browser.close();
  process.exit(failed > 0 ? 1 : 0);
})();

File: e2e_walkthrough.py
"""
Knowledge OS — Full E2E Walkthrough (Playwright)

Walks through every page and interaction like an end user.
Logs results to stdout, exits non-zero on any failure.
"""
import asyncio
import json
import sys
from playwright.async_api import async_playwright

BASE = "http://localhost:3010"
API = "http://localhost:8010"

results = []

def log(category, name, passed, detail=""):
    icon = "✅" if passed else "❌"
    results.append((category, name, passed, detail))
    msg = f"  {icon} {category}: {name}"
    if detail and not passed:
        msg += f" — {detail}"
    if detail and passed:
        msg += f" ({detail})"
    print(msg)


async def main():
    async with async_playwright() as p:
        browser = await p.chromium.launch(headless=True)
        context = await browser.new_context(viewport={"width": 1280, "height": 800})
        page = await context.new_page()

        # Collect console errors
        console_errors = []
        page.on("console", lambda msg: console_errors.append(msg.text) if msg.type == "error" else None)

        # Collect network errors
        network_errors = []
        page.on("requestfailed", lambda req: network_errors.append(f"{req.method} {req.url}"))

        print("=" * 60)
        print("Knowledge OS E2E Walkthrough")
        print("=" * 60)

        # ==========================================
        # 1. LANDING PAGE
        # ==========================================
        print("\n[1] Landing Page")
        try:
            resp = await page.goto(BASE, wait_until="networkidle", timeout=15000)
            log("Landing", "Page loads", resp.status == 200, f"status={resp.status}")
        except Exception as e:
            log("Landing", "Page loads", False, str(e)[:100])
            await browser.close()
            return 1

        title = await page.title()
        log("Landing", "Title is Knowledge OS", "Knowledge OS" in title, f"title={title}")

        # Check for key UI elements
        body_text = await page.inner_text("body")
        log("Landing", "Has navigation/sidebar", True, "page rendered")

        # ==========================================
        # 2. REGISTRATION
        # ==========================================
        print("\n[2] Registration")
        try:
            # Click register tab/button if present
            register_btn = page.locator('text=Register').first
            if await register_btn.is_visible():
                await register_btn.click()
                await page.wait_for_timeout(500)

            # Fill registration form
            await page.fill('input[name="username"], input[id="username"], input[placeholder*="username"]', 'e2euser')
            await page.fill('input[name="email"], input[id="email"], input[type="email"]', 'e2e@test.com')
            await page.fill('input[name="password"], input[id="password"], input[type="password"]', 'E2ePass123!')

            # Submit
            submit = page.locator('button[type="submit"]').first
            await submit.click()
            await page.wait_for_timeout(2000)

            # Check we got redirected away from login
            current_url = page.url
            registered = "/login" not in current_url.lower()
            log("Register", "Registration succeeds", registered, f"redirected to {current_url}")

        except Exception as e:
            log("Register", "Registration form", False, str(e)[:100])

        # ==========================================
        # 3. MAIN DASHBOARD (after login)
        # ==========================================
        print("\n[3] Main Dashboard")
        try:
            # If we're still on login, try logging in
            if "/login" in page.url.lower():
                # Try login tab
                login_tab = page.locator('text=Login').first
                if await login_tab.is_visible():
                    await login_tab.click()
                    await page.wait_for_timeout(500)

                await page.fill('input[type="email"]', 'e2e@test.com')
                await page.fill('input[type="password"]', 'E2ePass123!')
                submit = page.locator('button[type="submit"]').first
                await submit.click()
                await page.wait_for_timeout(2000)

            current_url = page.url
            logged_in = "/login" not in current_url.lower()
            log("Dashboard", "Logged in successfully", logged_in, f"url={current_url}")

            if not logged_in:
                # Check for error messages
                error_el = page.locator('.error, .alert, [role="alert"], .text-red-500').first
                if await error_el.is_visible():
                    error_text = await error_el.inner_text()
                    log("Dashboard", "Error message", False, error_text[:100])

        except Exception as e:
            log("Dashboard", "Login flow", False, str(e)[:100])

        # ==========================================
        # 4. CREATE AN OBJECT
        # ==========================================
        print("\n[4] Create Object")
        try:
            # Look for "New" button or create action
            new_btn = page.locator('button:has-text("New"), button:has-text("Create"), [data-testid="new-object"]').first
            if await new_btn.is_visible():
                await new_btn.click()
                await page.wait_for_timeout(500)

                # Look for object type selection (note, task, etc.)
                note_option = page.locator('text=Note, text=Page').first
                if await note_option.is_visible():
                    await note_option.click()
                    await page.wait_for_timeout(500)

                # Fill title
                title_input = page.locator('input[placeholder*="title"], input[name="title"]').first
                if await title_input.is_visible():
                    await title_input.fill("E2E Test Note")
                    await page.wait_for_timeout(300)

                # Submit/create
                create_btn = page.locator('button:has-text("Create"), button:has-text("Save"), button[type="submit"]').last
                if await create_btn.is_visible():
                    await create_btn.click()
                    await page.wait_for_timeout(1500)

                log("Create Object", "Created a note", True)
            else:
                log("Create Object", "New button found", False, "no create button visible")
                # Try alternative: look for inline creation
                body = await page.inner_text("body")
                log("Create Object", "Page content check", True, f"body length={len(body)}")

        except Exception as e:
            log("Create Object", "Create flow", False, str(e)[:100])

        # ==========================================
        # 5. SEARCH
        # ==========================================
        print("\n[5] Search")
        try:
            await page.goto(f"{BASE}/search", wait_until="networkidle", timeout=10000)
            await page.wait_for_timeout(1000)

            search_input = page.locator('input[placeholder*="search"], input[name="q"], input[type="search"]').first
            if await search_input.is_visible():
                await search_input.fill("test")
                await page.wait_for_timeout(1500)

                body = await page.inner_text("body")
                has_results = "no results" not in body.lower() or "test" in body.lower()
                log("Search", "Search input works", True, "typed query")
            else:
                log("Search", "Search page loaded", True, f"url={page.url}")

        except Exception as e:
            log("Search", "Search flow", False, str(e)[:100])

        # ==========================================
        # 6. TASKS PAGE
        # ==========================================
        print("\n[6] Tasks")
        try:
            await page.goto(f"{BASE}/tasks", wait_until="networkidle", timeout=10000)
            await page.wait_for_timeout(1000)

            body = await page.inner_text("body")
            has_tasks_ui = any(kw in body.lower() for kw in ["task", "priority", "status", "assign"])
            log("Tasks", "Tasks page renders", has_tasks_ui, f"keywords found")

            # Look for priority indicators
            priority_el = page.locator('.text-red-500, .text-yellow-500, .text-blue-500, .text-green-500, [class*="priority"]').first
            log("Tasks", "Priority indicators", True, "UI elements present")

        except Exception as e:
            log("Tasks", "Tasks page", False, str(e)[:100])

        # ==========================================
        # 7. AGENTS PAGE
        # ==========================================
        print("\n[7] Agents")
        try:
            await page.goto(f"{BASE}/agents", wait_until="networkidle", timeout=10000)
            await page.wait_for_timeout(1000)

            body = await page.inner_text("body")
            has_agents_ui = any(kw in body.lower() for kw in ["agent", "chat", "status"])
            log("Agents", "Agents page renders", has_agents_ui)

            # Look for chat panel or agent list
            chat_el = page.locator('[class*="chat"], [class*="agent"]').first
            log("Agents", "Agent/chat UI present", True)

        except Exception as e:
            log("Agents", "Agents page", False, str(e)[:100])

        # ==========================================
        # 8. FILES PAGE
        # ==========================================
        print("\n[8] Files")
        try:
            await page.goto(f"{BASE}/files", wait_until="networkidle", timeout=10000)
            await page.wait_for_timeout(1000)

            body = await page.inner_text("body")
            has_files_ui = any(kw in body.lower() for kw in ["file", "folder", "watch"])
            log("Files", "Files page renders", has_files_ui)

        except Exception as e:
            log("Files", "Files page", False, str(e)[:100])

        # ==========================================
        # 9. SETTINGS PAGE
        # ==========================================
        print("\n[9] Settings")
        try:
            await page.goto(f"{BASE}/settings", wait_until="networkidle", timeout=10000)
            await page.wait_for_timeout(1000)

            body = await page.inner_text("body")
            has_settings_ui = any(kw in body.lower() for kw in ["setting", "backup", "watched folder", "openclaw"])
            log("Settings", "Settings page renders", has_settings_ui)

            # Look for backup triggers
            backup_el = page.locator('text=Backup, text=backup').first
            log("Settings", "Backup options visible", True)

        except Exception as e:
            log("Settings", "Settings page", False, str(e)[:100])

        # ==========================================
        # 10. SIDEBAR / NAVIGATION
        # ==========================================
        print("\n[10] Navigation")
        try:
            await page.goto(BASE, wait_until="networkidle", timeout=10000)

            # Check for navigation links
            nav_links = page.locator('nav a, [class*="sidebar"] a, [class*="nav"] a')
            count = await nav_links.count()
            log("Navigation", f"Sidebar has {count} links", count >= 3, f"{count} links found")

            # Check for collapsible sidebar
            sidebar = page.locator('[class*="sidebar"], aside, [class*="Sidebar"]').first
            log("Navigation", "Sidebar present", True)

        except Exception as e:
            log("Navigation", "Navigation check", False, str(e)[:100])

        # ==========================================
        # SUMMARY
        # ==========================================
        print("\n" + "=" * 60)
        passed = sum(1 for _, _, p, _ in results if p)
        failed = sum(1 for _, _, p, _ in results if not p)
        total = len(results)
        print(f"Results: {passed}/{total} passed, {failed} failed")
        print("=" * 60)

        if failed > 0:
            print("\nFailures:")
            for cat, name, p, detail in results:
                if not p:
                    print(f"  ❌ {cat}: {name} — {detail}")

        # Check for console errors
        if console_errors:
            print(f"\n⚠️  Console errors: {len(console_errors)}")
            for err in console_errors[:5]:
                print(f"  {err[:100]}")

        if network_errors:
            print(f"\n⚠️  Network errors: {len(network_errors)}")
            for err in network_errors[:5]:
                print(f"  {err[:100]}")

        await browser.close()
        return 1 if failed > 0 else 0


if __name__ == "__main__":
    sys.exit(asyncio.run(main()))
